README
Copyright (C) 2014 by Marian Marinov

Linux-Unshare
==========================

  Linux::Unshare - Perl interface for Linux unshare system call.


This trivial module provides interface to the Linux unshare system call. It
also provides the CLONE_* constants that are used to specify which kind of
unsharing must be performed. Note that some of these are still not implemented
in the Linux kernel, and others are still experimental.

The unshare system call allows a process to 'unshare' part of the process
context which was originally shared using clone(2).

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module does not require any 3rd party libraries, but it does require your machine to have the <sched.h> for build.
And also it requires fairly recent kernel to work.

BUGS AND SUGGESTIONS
    contact me at hackman at cpan.org

REPOSITORY
    http://github.com/hackman/linux-unshare

COPYRIGHT AND LICENCE

Copyright (C) 2009 by Boris Sukholitko
Copyright (C) 2014 by Marian Marinov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Linux::Unshare.

1.00  Tue Nov  1 05:20:19 2016
    - Added CLONE_NEWCGROUP flag

0.05  
    - Added CLONE_CONATAINER and CLONE_NEWUSER
	- Added tests for CLONE_NEWUSER and CLONE_CONTAINER
	- META.yml: Added repository and lincense resources to the meta description
	- Makefile.PL: Added CLONE_NEWUSER
	- Unshare.xs: Added CLONE_CONTAINER definition, which include CLONE_NEWNS, CLONE_NEWIPC, CLONE_NEWUTS, CLONE_NEWNET, CLONE_NEWUSER, CLONE_NEWPID

0.04  Mon Jul 28 13:11:06 2014
    - Added Marian as co-author of the module.
	- Updated the README to include Marian
	- Copyright updated to include Marian
	- Added script for building the archive out of the current dir
	- Added script to easily bump the version of the package
    - README: Added the copyright notice to the top of the file as it is required by the perl distribution standard
	- README: Removed the version from this file
0.03  Thu Jul 24 08:08:58 2014
    - Added tests for CLONE_FILES, CLONE_NEWUTS, CLONE_SYSVSEM, CLONE_NEWIPC, CLONE_NEWNET



Makefile.PL
die "OS Unsupported" if ($^O !~ m#(?i)Linux#);

use 5.010000;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME              => 'Linux::Unshare',
    VERSION_FROM      => 'lib/Linux/Unshare.pm',
    PREREQ_PM         => {},
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Linux/Unshare.pm',
       AUTHOR         => [ 'Boris Sukholitko <boris@>', 'Marian Marinov <HackMan>' ]) : ()),
    LIBS              => [''],
    DEFINE            => '',
    INC               => '-I.',
);
if (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw(CLONE_THREAD CLONE_FS CLONE_NEWNS CLONE_SIGHAND CLONE_VM
	  CLONE_FILES CLONE_SYSVSEM CLONE_NEWUTS CLONE_NEWIPC CLONE_NEWNET));
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Linux::Unshare',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc'
                                  );
} else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


lib/Linux/Unshare.pm
package Linux::Unshare;

use 5.010000;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# This allows declaration	use Linux::Unshare ':all';
our %EXPORT_TAGS = ( 'clone' => [ qw(
	CLONE_THREAD CLONE_FS CLONE_NEWNS CLONE_SIGHAND CLONE_VM CLONE_FILES CLONE_SYSVSEM
	CLONE_CONTAINER CLONE_NEWNS CLONE_NEWUTS CLONE_NEWIPC CLONE_NEWNET CLONE_NEWPID
	CLONE_NEWUSER CLONE_NEWCGROUP
) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'clone'} }, qw(unshare unshare_ns) );
our @EXPORT = qw( );

our $VERSION = '1.00';

sub AUTOLOAD {
    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&Linux::Unshare::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	*$AUTOLOAD = sub { $val };
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('Linux::Unshare', $VERSION);

sub unshare_ns { return unshare(0x20000) ? 0 : -1; }

1;
__END__
=head1 NAME

Linux::Unshare - Perl interface for Linux unshare system call.

=head1 SYNOPSIS

  use Linux::Unshare qw(unshare :clone);

  # as root ... 
  unshare(CLONE_NEWNS)
  # now your mounts will become private

  unshare(CLONE_NEWNET)
  # get a separate network namespace


=head1 DESCRIPTION

This trivial module provides interface to the Linux unshare system call. It
also provides the CLONE_* constants that are used to specify which kind of
unsharing must be performed. Note that some of these are still not implemented
in the Linux kernel, and others are still experimental.

The unshare system call allows a process to 'unshare' part of the process
context which was originally shared using clone(2).

=head1 SEE ALSO

unshare(2) Linux man page.

=head1 AUTHOR

Boris Sukholitko, E<lt>boriss@gmail.comE<gt>
Marian Marinov, E<lt>hackman@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Boris Sukholitko
Copyright (C) 2014 by Marian Marinov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut
